File: DynamicPrograming/go/max_path_sum.go
package dynamicprogramming

// MaxPathSum returns the maximum path sum from top-left to bottom-right moving only down or right.
func MaxPathSum(grid [][]int) int {
	rows := len(grid)
	if rows == 0 {
		return 0
	}
	cols := len(grid[0])
	memo := make(map[[2]int]int)

	var dfs func(int, int) int
	dfs = func(r, c int) int {
		if r == rows || c == cols {
			return minInt
		}
		if r == rows-1 && c == cols-1 {
			return grid[r][c]
		}
		key := [2]int{r, c}
		if val, ok := memo[key]; ok {
			return val
		}
		down := dfs(r+1, c)
		right := dfs(r, c+1)
		best := down
		if right > best {
			best = right
		}
		memo[key] = grid[r][c] + best
		return memo[key]
	}

	return dfs(0, 0)
}

File: DynamicPrograming/php/maxPathSum.php
<?php

declare(strict_types=1);

/** @param int[][] $grid */
function maxPathSum(array $grid): int
{
    $rows = count($grid);
    $cols = $rows ? count($grid[0]) : 0;
    $memo = [];

    $dfs = function (int $r, int $c) use (&$dfs, &$memo, $grid, $rows, $cols): int {
        if ($r === $rows || $c === $cols) {
            return PHP_INT_MIN;
        }
        if ($r === $rows - 1 && $c === $cols - 1) {
            return $grid[$r][$c];
        }
        $key = $r . ',' . $c;
        if (array_key_exists($key, $memo)) {
            return $memo[$key];
        }
        return $memo[$key] = $grid[$r][$c] + max($dfs($r + 1, $c), $dfs($r, $c + 1));
    };

    return $dfs(0, 0);
}
